Rectangle/WindowManager.swift

```diff
@@ -98,13 +99,16 @@ class WindowManager {
         }
 
         let resultingRect = frontmostWindowElement.rectOfElement()
-        windowHistory.lastRectangleActions[windowId] = RectangleAction(action: calcResult.resultingAction, rect: resultingRect)
+        windowHistory.lastRectangleActions[windowId] = RectangleAction(action: calcResult.resultingAction, rect: resultingRect, context: calcResult.context)
     }
 }
 
+typealias ActionContext = [String: Any]
```

Comment:
Let's see, where to start on this. Swift is a strongly typed language with some pretty useful object oriented concepts built in. By creating a type aliased dictionary with values any type, you're basically building an object that doesn't make use of any of Swift's strongly typed or object oriented (and potentially protocol-based) concepts. This typically makes for code that is unnecessarily difficult to read and maintain. By using the `Any` keyword, any time we want to obtain a value out of that dictionary, we will have to type check/cast it into what we expect it to be. As you could imagine, `Any` is something I always avoid because I almost always want typed objects if I can help it. Instead of a type aliased dictionary, I think what we really want is either a struct or a class that we can pass around. Lucky for us, we already have the `RectangleAction` below. 

Instead of a context object, let's add what we really want to keep track of straight to that struct, since it's already being passed into a `windowCalculation`. What do we want to keep track of? Looks like you built the SupportedPositions enum. Going with an enum for this approach is a good choice, and it's how I would have done it as well. However, let's take a step back and see if there's something better we can keep state of. Perhaps we can keep state without having to return anything extra from the window calculation. This would simplify our code a good bit, as there's not as many pieces scattered around that are giving us this state stored in the last window action. 

My thought here is that instead of keeping track of the SupportedPosition that is determined by the window calculation, we would rather keep track of a "Repeated Execution Count" of the window action. Inside of the WindowManager is where most all of our important logic is stored, and this is something that we can viably keep track of here. If we track this count as part of our `RectangleAction` and pass it along to the `calculateRect` in each `WindowCalculation`, then we can just as easily tell which calculation to do - 1/3, 2/3, etc. Also, we're not binding WindowActions to the enumeration. This is important because I plan on implementing next/prev thirds as part of the first third or last third actions. By passing each WindowCalculation a count instead of an enum, each action doesn't need to be aware of all the other cases in that enum - they have the count and they can `switch` on it to choose what to do.

I'll stop my review here for now - since changes from this comment would affect all the other changes that you have currently in the PR. If you want to keep working on this and have the time, by all means, go for it and I'll keep reviewing. If not - that's ok, too! I don't mind rolling it out.

Reply:
Thanks for your feedback/input, really appreciate you taking the time to review and respond.

The reason for my choice of a context with `Any` stems back from C/C++ days I guess, where a `void**` was passed into your method allowing your method store _it’s_ state in how _it_ saw fit, as that method was only going to be only thing that will be accessing it. Which absolutely agree, isn’t very type-safe.

I do like the idea of just keeping an execution count that is passed around. It's a lot cleaner and simpler, I'm annoyed that I didn't see, I think perhaps I was focus on the concept of a ‘context’/‘state’ driven solution. (I think I just got carried away with some of Swift's other cool features, multiple return values, etc.) 

Looking at it with fresh eyes I really don't want this to be merged in in its current state. It needs to be refactored…